core/src/kvs/ds.rs

```diff
@@ -920,6 +1005,67 @@ impl Datastore {
 		Ok(())
 	}
 
+	/// Construct a document from a Change Feed mutation
+	/// This is required to perform document operations such as live query notifications
+	fn construct_document(mutation: &TableMutation) -> Option<Document> {
+		match mutation {
+			TableMutation::Set(a, b) => {
+				let doc = Document::new(None, Some(a), None, b, Workable::Normal);
+				Some(doc)
+			}
+			TableMutation::Del(a) => {
+				let doc = Document::new(None, Some(a), None, &Value::None, Workable::Normal);
+				Some(doc)
+			}
+			TableMutation::Def(_) => None,
+			TableMutation::SetPrevious(id, _old, new) => {
+				let doc = Document::new(None, Some(id), None, new, Workable::Normal);
+				// TODO set previous value
+				Some(doc)
+			}
+		}
+	}
+
+	/// Add live queries to track on the datastore
+	/// These get polled by the change feed tick
+	pub(crate) async fn track_live_queries(&self, lqs: &Vec<TrackedResult>) -> Result<(), Error> {
+		// Lock the local live queries
+		let mut lq_map = self.local_live_queries.write().await;
+		let mut cf_watermarks = self.cf_watermarks.write().await;
+		for lq in lqs {
+			match lq {
+				TrackedResult::LiveQuery(lq) => {
+					let lq_index_key: LqIndexKey = lq.as_key();
+					let m = lq_map.get_mut(&lq_index_key);
+					match m {
+						Some(lq_index_value) => lq_index_value.push(lq.as_value()),
+						None => {
+							let lq_vec = vec![lq.as_value()];
+							lq_map.insert(lq_index_key.clone(), lq_vec);
+						}
+					}
+					let selector = lq_index_key.selector;
+					let mut_watermark = cf_watermarks.get_mut(&selector);
+					match mut_watermark {
+						Some(_vs) => {
+							// We don't have a versionstamp, and even if we did - the watermarks are only for consumed events
+							// So we don't want to overwrite and leave that to the cf heartbeat
+						}
+						None => {
+							// TODO: (phughk) - read watermark for catchup
+							// We insert the current watermark.
+							cf_watermarks.insert(selector, Versionstamp::default());
+						}
+					}
```

Comment:
You can avoid an extra lookup here by using entry avoiding the second lookup in insert.
This can also be done with `lq_map` but this would require always cloning the `lq_index_key` which might be more expensive than doing a second clone so I am only suggesting this change here.
```suggestion
					// If we don't have a versionstamp, and even if we did - the watermarks are only for consumed events
					// So we don't want to overwrite and leave that to the cf heartbeat
					//
					// TODO: (phughk) - read watermark for catchup
					// We insert the current watermark.
					cf_watermarks.entry(selector).or_insert_with(Versionstamp::default);
```

Reply:
Addressed thanks!